startup/10-detectors.py

```diff
@@ -96,102 +98,149 @@ class Encoder(Device):
+            #self.filepath.put(self._ioc_full_path)
+
+            self._resource_uid = str(uuid.uuid4())
+            resource = {'spec': 'PIZZABOX_ENC_FILE_TXT',
+                        'root': DIRECTORY,
+                        'resource_path': full_path,
+                        'resource_kwargs': {'chunk_size': self.chunk_size},
+                        'path_semantics': os.name,
+                        'uid': self._resource_uid}
+            self._asset_docs_cache.append(('resource', resource))
+            self._datum_counter = itertools.count()
 
             super().stage()
-        else:
-            print("encoder was not staged : {}".format(self.name))
+            print('Staging of {} complete'.format(self.name))
 
     def unstage(self):
         if(self.connected):
             set_and_wait(self.ignore_sel, 1)
+            self._datum_counter = None
             return super().unstage()
 
     def kickoff(self):
-        "Start writing data into the file."
-
         print('kickoff', self.name)
         self._ready_to_collect = True
+        "Start writing data into the file."
 
-        # set_and_wait(self.ignore_sel, 0)
-        st = self.ignore_sel.set(0)
+        set_and_wait(self.ignore_sel, 0)
 
         # Return a 'status object' that immediately reports we are 'done' ---
         # ready to collect at any time.
-        # return NullStatus()
-        return st
+        return NullStatus()
 
     def complete(self):
-        print('complete', self.name, '| filepath', self._full_path)
+        print('storing', self.name, 'in', self._full_path)
         if not self._ready_to_collect:
             raise RuntimeError("must called kickoff() method before calling complete()")
         # Stop adding new data to the file.
-        #set_and_wait(self.ignore_sel, 1)
-        st = self.ignore_sel.set(1)
+        set_and_wait(self.ignore_sel, 1)
         #while not os.path.isfile(self._full_path):
         #    ttime.sleep(.1)
-        #return NullStatus()
-        return st
+
+        # FIXME: beam line disaster fix.
+        # Let's move the file to the correct place
+        #workstation_file_root = '/mnt/xf08ida-ioc1/'
+        #workstation_full_path = os.path.join(workstation_file_root, self._filename)
+        #print('Moving file from {} to {}'.format(workstation_full_path, self._full_path))
+        #cp_stat = shutil.copy(workstation_full_path, self._full_path)
+
+
+        # HACK: Make datum documents here so that they are available for collect_asset_docs
+        # before collect() is called. May need changes to RE to do this properly. - Dan A.
+
+        self._datum_ids = []
+
+        datum_id = '{}/{}'.format(self._resource_uid,  next(self._datum_counter))
+        datum = {'resource': self._resource_uid,
+                 'datum_kwargs': {},
+                 'datum_id': datum_id}
+        self._asset_docs_cache.append(('datum', datum))
+
+        self._datum_ids.append(datum_id)
+
+        return NullStatus()
```

Comment:
These updates look familiar. Are they based on the changes from ISS? I remember @danielballan and I were working on something similar...